telemetry/src/main/java/datadog/telemetry/TelemetryServiceImpl.java

```diff
@@ -80,94 +47,178 @@ public boolean addConfiguration(Map<String, Object> configuration) {
+    RequestStatus status = sendRequest(request);
+
+    if (status == RequestStatus.SUCCESS) {
+      dependencies.removeAll(deps);
+    }
+
+    return status;
+  }
+
+  private RequestStatus sendMetrics(RequestBuilder requestBuilder) {
+    List<Metric> mtrs = new LinkedList<>(metrics);
+
+    Payload payload =
+        new GenerateMetrics()
+            .namespace("appsec")
+            .libLanguage("java")
+            .libVersion("0.0.0")
+            .series(mtrs)
+            .requestType(RequestType.GENERATE_METRICS);
+
+    Request request = requestBuilder.build(RequestType.GENERATE_METRICS, payload);
+    RequestStatus status = sendRequest(request);
+
+    if (status == RequestStatus.SUCCESS) {
+      metrics.removeAll(mtrs);
     }
 
-    return queue;
+    return status;
   }
 
   @Override
-  public int getHeartbeatInterval() {
-    return heartbeatIntervalMs;
+  public RequestStatus sendTelemetry(RequestBuilder requestBuilder) {
+    if (requestBuilder == null) {
+      return RequestStatus.ENDPOINT_ERROR;
+    }
+
+    RequestStatus status;
+
+    if (!integrations.isEmpty()) {
+      status = sendIntegrations(requestBuilder);
+      if (status != RequestStatus.SUCCESS) {
+        return status;
+      }
+    }
+
+    if (!dependencies.isEmpty()) {
+      status = sendDependencies(requestBuilder);
+      if (status != RequestStatus.SUCCESS) {
+        return status;
+      }
+    }
+
+    if (!metrics.isEmpty()) {
+      status = sendMetrics(requestBuilder);
+      if (status != RequestStatus.SUCCESS) {
+        return status;
+      }
+    }
+
+    return sendHeartbeat(requestBuilder);
   }
 
-  private static <T> List<T> drainOrNull(BlockingQueue<T> srcQueue) {
-    return drainOrDefault(srcQueue, null);
+  @Override
+  public RequestStatus sendAppClosing(RequestBuilder requestBuilder) {
+    Request request = requestBuilder.build(RequestType.APP_CLOSING);
+    return sendRequest(request);
   }
 
-  private static <T> List<T> drainOrEmpty(BlockingQueue<T> srcQueue) {
-    return drainOrDefault(srcQueue, Collections.<T>emptyList());
+  public RequestStatus sendHeartbeat(RequestBuilder requestBuilder) {
+    Request request = requestBuilder.build(RequestType.APP_HEARTBEAT);
+    return sendRequest(request);
   }
 
-  private static <T> List<T> drainOrDefault(BlockingQueue<T> srcQueue, List<T> defaultList) {
-    List<T> list = new LinkedList<>();
-    int drained = srcQueue.drainTo(list);
-    if (drained > 0) {
-      return list;
+  private RequestStatus sendRequest(Request request) {
+    try (Response response = okHttpClient.newCall(request).execute()) {
+      switch (response.code()) {
+        case 202:
+          return RequestStatus.SUCCESS;
+
+        case 404:
+          return RequestStatus.ENDPOINT_ERROR;
+
+        default:
+          return RequestStatus.HTTP_ERROR;
+      }
+    } catch (IOException e) {
+      log.warn("IOException on HTTP request to Telemetry Intake Service", e);
```

Comment:
The stacktrace is rarely useful in this case, maybe we should stick to this?

```suggestion
      log.warn("IOException on HTTP request to Telemetry Intake Service: {}", e.getMessage());
```